vsphere/internal/helper/envbrowse/environment_browser_helper.go

```diff
@@ -0,0 +1,116 @@
+)
+
+// EnvironmentBrowser is a higher-level interface to a specific object's
+// environment browser.
+//
+// This essentially fills he role of such functionality lacking in govmomi at
+// this point in time and may serve as the basis for a respective PR at a later
+// point in time.
+type EnvironmentBrowser struct {
+	object.Common
+}
+
+// NewEnvironmentBrowser initializes a new EnvironmentBrowser based off the
+// supplied managed object reference.
+func NewEnvironmentBrowser(c *vim25.Client, ref types.ManagedObjectReference) *EnvironmentBrowser {
+	return &EnvironmentBrowser{
+		Common: object.NewCommon(c, ref),
+	}
+}
+
+// DefaultDevices loads a satisfactory default device list for the optionally
+// supplied guest ID list, host, and descriptor key. The result is returned as
+// a higher-level VirtualDeviceList object. This can be used as an initial
+// VirtualDeviceList when building a device list and VirtualDeviceConfigSpec
+// list for new virtual machines.
+//
+// Appropriate options for key can be loaded by running
+// QueryConfigOptionDescriptor, which will return a list of
+// VirtualMachineConfigOptionDescriptor which will contain the appropriate key
+// to use under key. If no key is supplied, the results generally reflect the
+// most recent VM hardware version.
+func (b *EnvironmentBrowser) DefaultDevices(ctx context.Context, guests []string, key string, host *object.HostSystem) (object.VirtualDeviceList, error) {
+	var eb mo.EnvironmentBrowser
+
+	err := b.Properties(ctx, b.Reference(), nil, &eb)
+	if err != nil {
+		return nil, err
+	}
+
+	req := types.QueryConfigOptionEx{
+		This: b.Reference(),
+		Spec: &types.EnvironmentBrowserConfigOptionQuerySpec{
+			Key:     key,
+			GuestId: guests,
+		},
+	}
+	if host != nil {
+		ref := host.Reference()
+		req.Spec.Host = &ref
+	}
+	res, err := methods.QueryConfigOptionEx(ctx, b.Client(), &req)
+	if err != nil {
+		return nil, err
+	}
+	if res.Returnval == nil {
+		return nil, errors.New("no config options were found for the supplied criteria")
+	}
+	return object.VirtualDeviceList(res.Returnval.DefaultDevice), nil
+}
+
+// OSFamily fetches the operating system family for the supplied guest ID.
+func (b *EnvironmentBrowser) OSFamily(ctx context.Context, guest string) (string, error) {
+	var eb mo.EnvironmentBrowser
+
+	err := b.Properties(ctx, b.Reference(), nil, &eb)
+	if err != nil {
+		return "", err
+	}
+
+	req := types.QueryConfigOptionEx{
+		This: b.Reference(),
+		Spec: &types.EnvironmentBrowserConfigOptionQuerySpec{
+			GuestId: []string{guest},
+		},
+	}
+	res, err := methods.QueryConfigOptionEx(ctx, b.Client(), &req)
+	if err != nil {
+		return "", err
+	}
+	if res.Returnval == nil {
+		return "", errors.New("no config options were found for the supplied criteria")
+	}
+	if len(res.Returnval.GuestOSDescriptor) < 1 {
+		return "", errors.New("no guest OS descriptors were found")
+	}
+	family := res.Returnval.GuestOSDescriptor[0].Family
```

Comment:
Do we need to check if this returns more than one `GuestOSDescriptor` or is the first one the one we are looking for?

Reply:
According to `QueryConfigOptionEx` in [`EnvironmentBrowser`](https://pubs.vmware.com/vsphere-6-5/index.jsp?topic=%2Fcom.vmware.wssdk.apiref.doc%2Fvim.EnvironmentBrowser.html), it looks like `GuestOSDescriptor` is actually filtered against the guest IDs provided. So unless there are two entries specifically matching the guest ID, this should never happen. Added an assertion on this though.